src/Agent.Worker/ContainerOperationProvider.cs

```diff
@@ -142,6 +147,74 @@ public async Task StopContainersAsync(IExecutionContext executionContext, object
             await RemoveContainerNetworkAsync(executionContext, _containerNetwork);
         }
 
+        private async Task<string> GetMSILoginToken(CancellationToken cancellationToken)
+        {
+            // Check environment variable for debugging
+            var envVar = System.Environment.GetEnvironmentVariable("DEBUG_MSI_LOGIN_INFO");
+            // Future: Set this client id. This is the MSI client ID.
+            TokenCredential credential = new ManagedIdentityCredential(clientId: null);
+            if (envVar == "1")
+            {
+                // Future: Set this client id. This is the MSI client ID.
+                credential = new ChainedTokenCredential(new ManagedIdentityCredential(clientId: null), new VisualStudioCredential(), new AzureCliCredential());
+            }
+            var accessToken = await credential.GetTokenAsync(new TokenRequestContext(new[] { "https://management.core.windows.net/" }), cancellationToken);
+            // To print the token, you can convert it to string 
+            return accessToken.Token.ToString();
+        }
+
+        private async Task<string> GetACRPasswordFromAADToken(string AADToken, string tenantId, string registryServer, int retryCount, int timeToWait, CancellationToken cancellationToken)
+        {
+            Uri url = new Uri(registryServer + "/oauth2/exchange");
+            const int retryLimit = 5;
+            using var httpClientHandler = HostContext.CreateHttpClientHandler();
+            using HttpClient httpClient = new HttpClient(httpClientHandler);
+            httpClient.DefaultRequestHeaders.Add("Content-Type", "application/x-www-form-urlencoded");
+            using StringContent stringContent = new StringContent("grant_type=access_token&service=" + registryServer[8..] + "&tenant=" + tenantId + "&access_token=" + AADToken);
+            string ACRpassword = string.Empty;
+            int waitedTime = 0;
+            do
+            {
+                try
+                {
+                    var response = await httpClient.PostAsync(url, stringContent, cancellationToken).ConfigureAwait(false);
```

Comment:
Can you use https://docs.microsoft.com/en-us/dotnet/api/system.net.http.formurlencodedcontent?view=net-6.0